### DISCUSSION TITLE
Support for sharing secrets between jobs

### DISCUSSION THREAD

**Discussion by danielmarbach** [community/community#13082] [Actions]
https://github.com/actions/runner/issues/1498#issuecomment-1066836352

> To give you some more insights. Normally we would want to do everything we can as part of one job. Yet in this specific case we are setting a server cluster that requires a few nodes to work. That is an expensive operations and we are working against a limited set of resources. By having the setup of the cluster in the same matrix build job we would then setup the cluster per matrix which would quickly lead to resource exhaustion. 
> 
> To accomodate that we have created a setup job that creates the cluster. The matrix builds wait for the setup job to be completed. Then there is also a cleanup job that runs after all the matrix builds have run or things have failed. Both the matrix jobs as well as the cleanup need information about how and where to access the cluster in order to be able to connect to it and eventually destroy it again after the run. We want to avoid having this information to be leaked. Hence we were hoping to "just mask the dynamic secrets" and then share them with jobs. 
> 
> I can understand though the design and architectural reasons why that is not allowed (or supported). It just means for any such a scenario you are basically forced to reinvent the wheel like we did.
> 
> here is the encryption step we ended up using
> 
> https://github.com/Particular/NServiceBus.RavenDB/blob/74f4ee2c1ad17327e8da523667a2b9f8aa323df0/.github/workflows/ci.yml#L125-L129
> 
> here the artifact upload
> 
> https://github.com/Particular/NServiceBus.RavenDB/blob/74f4ee2c1ad17327e8da523667a2b9f8aa323df0/.github/workflows/ci.yml#L130-L137
> 
> then the explicit download, decrypt steps
> 
> https://github.com/Particular/NServiceBus.RavenDB/blob/74f4ee2c1ad17327e8da523667a2b9f8aa323df0/.github/workflows/ci.yml#L168-L184

**Comment by ethomson** [comment]
Thanks @danielmarbach. I appreciate the helpful example - this isn't something that we support today but your use case makes a lot of sense and is something that we'll consider in the future. I've added this to the backlog.

Thanks again for the helpful feedback.

**Reply by Alleged627** [reply]
Can you explain what they are asking me to do and ami going to get in trouble

**Comment by KevinMSampson** [comment]
To add to this we should have a way to target jobs that can use secrets. I don't want to have someone print or use a secret from a called workflow job in a caller workflow step afterwards.

**Reply by danielmarbach** [reply]
Agreed

**Comment by mike-kaminski** [comment]
Could this be handled by having the setup job run an API operation to either create a temporary repo secret using an easily constructed naming convention, or simply update a statically assigned secret each run that can be leveraged by subsequent jobs? You could add the temporary secret's ID as an output that the consuming job(s) reference for the value (the secret name/ID) and then utilizes as appropriate? It'd be less "clunky" and could also be used for other workflows, or if it's an organization level secret instead, be used by other repos in the org. Perhaps as a method to share that temporary infrastructure?

**Comment by jnahmias** [comment]
I've run into this exact requirement myself, see https://github.com/orgs/community/discussions/29880 for more details on my use-case.
@ethomson - any update on where this is on the roadmap? Having a way to share ephemeral random tokens within a workflow would be really useful.

**Comment by JeffreyArt1** [comment]
Any updates on this?

![spying](https://media.giphy.com/media/zQc8STzaOlJ3q/giphy.gif)

**Comment by danielmarbach** [comment]
FYI in the meantime we moved away from using multiple jobs towards having custom actions with setup and teardown. I will try to update the links in this issue to point to the previous version

**Reply by Alleged627** [reply]
What does any of this mean I feel like all of this is gonna get me in trouble and honestly I don’t know what to do

**Reply by mikocot** [reply]
this is not really a solution or even a workaround, the reason to divide workflow in jobs is not only for readability (which could be indeed achieved by compositw actions instead) but to have independent blocks that can be run or not, that can fail or not and therefore control the flow, and finally to also run matrices.

If you have a matrix that is going to trigger 10 runs, you don't want, or often cannot, run all the previous actions that led to it in each of those runs.

**Reply by danielmarbach** [reply]
That is true. My FYI was never intended to hint or say what we did was a solution or workaround. I just wanted to give a heads up our reality has changed. The problem as you rightfully pointed out still exists.

**Comment by jonashartwig** [comment]
Hi, we at Telia Company have a similar requirement that grows out of company directives. They require easy modular reusable workflows and proper secret handling. For secrets Vault is used.

So here is what we struggle with:

We have a re-usable workflow that deploys any helm chart from either the git repo or a remote helm repo. For lots of use cases that re-usable workflow must be provided with secrets that are external to that workflow. So what we tried to do is this:

```yaml
name: Build and deploy

on:
  push:
    branches:
      - "*"
    tags:
      - "v*"

jobs:
  build:
    uses: telia-company/devx-cicd-workflows-and-actions/.github/workflows/build-helm.yml@main
    with:
      squad: "popsicle"
      chartValueFiles: values/sit.yaml
    # these secrets are ok, and can be loaded inside the re-usable workflow, they are always the same
    secrets:
      repositoryUsername: ${{ secrets.ARCUS_JFROG_WRITER_USERNAME }}
      repositoryPassword: ${{ secrets.ARCUS_JFROG_WRITER_PASSWORD }}
  vault:
    runs-on: ["telia-managed"]
    steps:
      - name: Import Secrets
        uses: hashicorp/vault-action@v2.4.0
        id: vault
        with:
          url: https://some.url
          namespace: Hid100007639
          # these must be on the repository anyway
          token: ${{ secrets.ARCUS_VAULT_TOKEN }}
          # here starts the problem, these secrets must come into the next job
          secrets: |
            kv/data/github secret1;
            kv/data/github secret2
  deploy-sit:
    needs: [ "vault" ]
    uses: telia-company/devx-cicd-workflows-and-actions/.github/workflows/deploy-helm.yml@main
    with:
      squad: popsicle
      namespace: infra
      chartValueFiles: values/sit.yaml
      releaseName: arcus
    secrets:
      # here is the problem, this is not easy possible, outputs are rightfully redacted as this are secrets and we do not have another way
      chartValues: |
        secret1: ${{ needs.vault.secrets.secret1 }}
        secret2: ${{ needs.vault.secrets.secret2 }}
      k8sConfig: this should also come from vault
```

What we want is a way to simply and safe share secrets between jobs in one run. It would be great if a workflow like outputs could get secrets like feature so those are available through needs context. Or any other solution of course :)

regards

**Reply by romain-cambonie** [reply]
@jonashartwig  
https://nitratine.net/blog/post/how-to-pass-secrets-between-runners-in-github-actions/

TL;DR; you need to encrypt / decrypt the secret.

Using this article as base I successfully passed secrets between jobs.
 
An example in a workflow that pass encrypted AWS secrets between jobs : https://github.com/romain-cambonie/serenity-workflow-dev/blob/master/.github/workflows/_add-aws-credentials-to-workspace.terraform.reusable.yml

**Reply by jonashartwig** [reply]
Hi, thanks. We have read that article multiple times and considered doing it that way. However we think the setup and maintenance of this is too much.

**Reply by h-r-k-matsumoto** [reply]
I faced same issue when using vault-action.
I resolved it by distributing the secrets of Vault to secret in a repository like below.
https://github.com/hashicorp/vault-action/compare/main...h-r-k-matsumoto:vault-action:feature/gh-secrets

I know this is not desirable.
It cause to persist secret unnecessarily.
However, it is possible to centrally manage secret with Vault.

**Comment by Alleged627** [comment]
Hi this is Keli what is it you would like me to do I want to help however I can I apologize for my lack of knowledge I am trying to learn more.

**Reply by tvquizphd** [reply]
Hi @Alleged627. Are you directing your comments to @danielmarbach ? This is a world-wide discussion.

I'm running into the same issue on my own project. GitHub Actions allows workflows with multiple jobs.
It seems that jobs cannot share "secrets" among themselves.
So, any info shared between jobs is publicly visible.

The main workaround now is to use encryption before sharing info between jobs.

**Reply by kesireddymanikanta1** [reply]
secrets need to encoded in first job and then later decoded in another job

**Comment by rdhar** [comment]
###### BACKGROUND CONTEXT

Although there are now docs on [masking and passing secrets between jobs or workflows](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#example-masking-and-passing-a-secret-between-jobs-or-workflows) (as @romain-cambonie [linked](https://github.com/orgs/community/discussions/13082#discussioncomment-3922380)), it didn't meet requirements like:

- Storing and retrieving from a secret store seems a tad **over-the-top** (as @jonashartwig [mentioned](https://github.com/orgs/community/discussions/13082#discussioncomment-3933094)).
  - The same goes for encrypted artifact upload/download.
- For ephemeral, temporary [OIDC](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-cloud-providers) tokens, it's **not feasible to dynamically store/retrieve** those secrets (as @jnahmias [raised](https://github.com/orgs/community/discussions/13082#discussioncomment-3422634)).

###### PROBLEM STATEMENTS

With these factors in mind, the main blockers are:

- Per docs, to **prevent accidental disclosures**, GitHub Actions [redacts any configured secrets as well as common encodings of those values](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-secrets), like Base64.
- For dynamically-generated secrets, those have to **remain masked from the caller workflow** to the reusable workflow to avoid exposing their their values in the workflow logs.

###### PROPOSED SOLUTION

1. Before going to output secrets from the caller workflow, Base64 encode the values **twice**.
   1. This enables GitHub Actions to output the secrets without exposing them.
1. Per docs, pass the encoded values to the reusable workflow as [secret inputs](https://docs.github.com/en/actions/using-workflows/reusing-workflows#passing-inputs-and-secrets-to-a-reusable-workflow).
   1. This enables the reusable workflow accept and treat the encoded values as secrets.
1. In the reusable workflow, decode the values from Base64 **twice** before masking them.
   1. This ensures that the secrets remain masked in the workflow logs throughout.
   1. They can now be populated as environment variables for use in the subsequent steps of the reusable workflow.

###### WORKING EXAMPLE

These methods are sourced from [DevSecTop/TF-via-PR](https://github.com/devsectop/tf-via-pr/blob/main/.github/workflows/caller_aws.yml) ([permalink](https://github.com/devsectop/tf-via-pr/blob/7b251158cc2985139df074c9ed6738d93901495b/.github/workflows/caller_aws.yml)) repository, which hosts a reusable workflow to run Terraform commands via PR comments, like a CLI.

As a bonus, any number of secrets can be securely passed into the reusable workflow to be used as environment variables, for example. The repository also contains recent [GitHub Actions workflow runs](https://github.com/devsectop/tf-via-pr/actions) to verify that the secrets remain masked throughout.

```yml
# caller-workflow.yml

jobs:
  credentials:
    runs-on: ubuntu-latest
    outputs:
      CREDENTIAL1: ${{ steps.credentials.outputs.CREDENTIAL1 }}
      CREDENTIAL2: ${{ steps.credentials.outputs.CREDENTIAL2 }}
    steps:
      - name: Output encoded credentials
        id: credentials
        env:
          CREDENTIAL1: ${{ secrets.CREDENTIAL1 }}
          CREDENTIAL2: ${{ secrets.CREDENTIAL2 }}
        run: |
          echo "CREDENTIAL1=$(echo $CREDENTIAL1 | base64 -w0 | base64 -w0)" >> $GITHUB_OUTPUT
          echo "CREDENTIAL2=$(echo $CREDENTIAL2 | base64 -w0 | base64 -w0)" >> $GITHUB_OUTPUT

  reusable-workflow:
    needs: credentials
    uses: reusable-workflow.yml
    secrets:
      env_vars: |
        CREDENTIAL1=${{ needs.credentials.outputs.CREDENTIAL1 }}
        CREDENTIAL2=${{ needs.credentials.outputs.CREDENTIAL2 }}
```

```yml
# reusable-workflow.yml

on:
  workflow_call:
    secrets:
      env_vars:
        required: true
jobs:
  parse-credentials:
    runs-on: ubuntu-latest
    env:
      env_vars: ${{ secrets.env_vars }}
    steps:
      - name: Decode credentials as environment variables
        run: |
          for i in $env_vars; do
            i=$(echo $i | sed 's/=.*//g')=$(echo ${i#*=} | base64 -di | base64 -di)
            echo ::add-mask::${i#*=}
            printf '%s\n' "$i" >> $GITHUB_ENV
          done
      - name: Validate credentials
        run: |
          # Secrets are now available as masked environment variable.
          echo $CREDENTIAL1 # or ${{ env.CREDENTIAL1 }}
          echo $CREDENTIAL2 # or ${{ env.CREDENTIAL2 }}
```

Where:

- `base64 -w0`: Ensures that the encoded values are outputted in a single line, regardless of the length of the input.
- `base64 -di`: Decodes the values from Base64, ignoring any newlines from `echo`'d strings.
- `echo ${i#*=}`: Removes the key of the key-value pair using parameter expansion, leaving only the value.
- `sed 's/=.*//g'`: Removes the value of the key-value pair, leaving only the key.
- `echo ::add-mask::${i#*=}`: Masks the value of the key-value pair [without outputting](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#example-masking-a-generated-output-within-a-single-job).

###### FLAWS

- If the workflow is (re-)run in debug mode, then secrets from `credentials` job specifically are output in the workflow log.
- The reusable workflow needs to be configured with the steps under `parse-credentials` job.

**Reply by shinebayar-g** [reply]
By outputting your secrets in the `credentials` job you're logging your secret in the job output. (_Only in the debug run output._)

As a workaround you should encrypt your secrets in the first job and then decrypt in the second job. But this requires you have an ownership on both jobs which is not the case if you're using someone else's reusable workflow.

**Reply by rdhar** [reply]
Hey @shinebayar-g, I used the think the exact same as well, but the evidence demonstrates otherwise.

**With regards to exposing secrets in the `credentials` job**, I have used the same method in this [public repo's actions](https://github.com/devsectop/tf-via-pr/actions). You can review the workflow logs directly yourself: attached the output below where secrets are never exposed, in base64 encoding or otherwise. 

<details><summary>Toggle view of workflow log...</summary>

```shell
Complete job name: credentials
##[group]Run aws-actions/configure-aws-credentials@5fd3084fc36e372ff1fff382a39b10d03659f355
with:
 aws-region: us-east-1
 role-to-assume: ***
 audience: sts.amazonaws.com
##[endgroup]
(node:1704) NOTE: We are formalizing our plans to enter AWS SDK for JavaScript (v2) into maintenance mode in 2023.

Please migrate your code to use AWS SDK for JavaScript (v3).
For more information, check the migration guide at https://a.co/7PzMCcy
(Use `node --trace-warnings ...` to show where the warning was created)
##[group]Run echo "AWS_ACCESS_KEY_ID=$(echo $AWS_ACCESS_KEY_ID | base64 -w0 | base64 -w0)" >> $GITHUB_OUTPUT
[36;1mecho "AWS_ACCESS_KEY_ID=$(echo $AWS_ACCESS_KEY_ID | base64 -w0 | base64 -w0)" >> $GITHUB_OUTPUT[0m
[36;1mecho "AWS_SECRET_ACCESS_KEY=$(echo $AWS_SECRET_ACCESS_KEY | base64 -w0 | base64 -w0)" >> $GITHUB_OUTPUT[0m
[36;1mecho "AWS_SESSION_TOKEN=$(echo $AWS_SESSION_TOKEN | base64 -w0 | base64 -w0)" >> $GITHUB_OUTPUT[0m
shell: /usr/bin/bash -e {0}
env:
 AWS_DEFAULT_REGION: us-east-1
 AWS_REGION: us-east-1
 AWS_ACCESS_KEY_ID: ***
 AWS_SECRET_ACCESS_KEY: ***
 AWS_SESSION_TOKEN: ***
##[endgroup]
Post job cleanup.
Evaluate and set job outputs
Set output 'AWS_ACCESS_KEY_ID'
Set output 'AWS_SECRET_ACCESS_KEY'
Set output 'AWS_SESSION_TOKEN'
Cleaning up orphan processes
```
</details>

**With regards to exposing secrets in the _reusable workflow's_ `parse-credentials` job**, I have again used the same method in the [same public repo's actions](https://github.com/devsectop/tf-via-pr/actions). Since the credentials are passed in via `secrets:`, this is GitHub's official way of [passing secrets from the caller to the caller workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-inputs-and-secrets-in-a-reusable-workflow). Once again, you can review my workflow logs directly: attached the output where secrets are parsed without exposure.

<details><summary>Toggle view of workflow log...</summary>

```shell
##[group]Run echo PROMPT_MATRIX=$matrix >> $GITHUB_OUTPUT
[36;1mecho PROMPT_MATRIX=$matrix >> $GITHUB_OUTPUT[0m
[36;1mif [ -z "$env_vars" ]; then[0m
[36;1m  exit 0[0m
[36;1melse[0m
[36;1m  # Iterate over all key-value pairs passed into this workflow and[0m
[36;1m  # add them as environment variables for use in subsequent steps.[0m
[36;1m  # Keys prefixed with "BASE64_" will be decoded from base64 and the[0m
[36;1m  # prefix will be removed. E.g., "BASE64_KEY=value" becomes "KEY=value".[0m
[36;1m  # Keys prefixed with "CONFIG_TF_" will not be masked from logs to[0m
[36;1m  # avoid masking non-sensitive configuration variables of this workflow.[0m
[36;1m  for i in $env_vars; do[0m
[36;1m    if [[ $i == BASE64_* ]]; then i=$(echo $i | sed 's/^BASE64_//g' | sed 's/=.*//g')=$(echo ${i#*=} | base64 -di | base64 -di); fi[0m
[36;1m    if [[ $i != CONFIG_TF_* ]]; then echo ::add-mask::${i#*=}; fi[0m
[36;1m    printf '%s\n' $i >> $GITHUB_ENV[0m
[36;1m  done[0m
[36;1mfi[0m
shell: /usr/bin/bash -e {0}
env:
 TF_CLI_ARGS: -no-color
 TF_IN_AUTOMATION: true
 TF_INPUT: false
 env_vars: ***
```

</details>

Please feel free to let me know if I've missed some other detail as I have "skin in the game" by using this method to [pass AWS credentials in public workflows](https://github.com/devsectop/tf-via-pr/blob/main/.github/workflows/caller_aws.yml)!

**Reply by shinebayar-g** [reply]
Hey @rdhar thanks for letting me know the _missed_ detail. You are right. Job output is not logged. It just says `Set output <name>`. For the reusable workflow, input is received as `secrets:` thus it's hidden from the job input. It's a great workaround _when you have control_ of the reusable workflow.

**Reply by shinebayar-g** [reply]
Hey @rdhar , have you tried running the `credentials` job with `Enable debug logging` checked? I think that might show outputs.

**Reply by rdhar** [reply]
Hey @shinebayar-g, you're spot-on, (re-)running the workflow in debug mode does output secrets in the `credentials` job, where it's isolated. And some level of control is required over the reusable workflow in order to parse the encoded credentials from the caller workflow. 

I've now listed both of these flaws in the original post to ensure users are made aware.

**Reply by shinebayar-g** [reply]
Yeah, gotta make sure you're not accidentally leaking secrets in the debug run. I haven't found a way to check if the workflow is running with `Enable debug logging` enabled. _(`ACTIONS_RUNNER_DEBUG` variable mentioned on docs is apparently always empty unless you put some secret on the repo)_ If you ever happen to find that please let me know :)

**Reply by awgeorge** [reply]
Try ${{ runner.debug }}

**Comment by shinebayar-g** [comment]
> Outputs containing secrets are redacted on the runner and not sent to GitHub Actions.

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idoutputs

This makes reusable workflows useless when we want to download some secrets from somewhere and pass it to reusable workflows.
Whole encrypt/decrypt thing doesn't work when we want to use someone else's reusable workflow which we cannot control.

**Comment by ChrisPage-AT** [comment]
It would be really fantastic if Github supported secrets in "outputs" variables for jobs.  I wouldn't care if it's a separate section named "secret outputs" or keep things in outputs and have the ability to add a "secret: true" flag.  In workflows with multiple jobs, it would be really nice to only have to make one call to secrets providers and then be able to use that secret throughout the whole workflow via output.

Separate sections example:
```
jobs:
  get-workflow-values:
    runs-on: linux-small
    name: Get Workflow Values
    outputs:
     notSecretValue: ${{ steps.get-not-secret-value.outputs.value }}
    secret-outputs:
     secretValue: ${{ steps.get-secret-value.outputs.value }}
```

Flag example:
```
jobs:
  get-workflow-values:
    runs-on: linux-small
    name: Get Workflow Values
    outputs:
     notSecretValue: ${{ steps.get-not-secret-value.outputs.value }}
     secretValue: 
      description: "This is a secret value for things"
      value: ${{ steps.get-secret-value.outputs.value }}
      secret: true
```

**Reply by jrbe228** [reply]
@ChrisPage-AT - does this [action](https://github.com/cloudposse/github-action-secret-outputs) address your use case?

**Reply by ChrisPage-AT** [reply]
I think that would work in my case, thank you.  It looks a little bit cumbersome, but any solution for this not built-in to github is going to be.

**Comment by CpuID** [comment]
The biggest benefit to this would be being able to use credentials generated in a prior job to access private container registries. These days short-lived credential generation using roles + OIDC etc are preferable to hardcoded secrets (eg. in Github Actions secrets in repo/org settings), as they eventually need rotation for compliance reasons etc.

This leads to using actions such as [`aws-actions/amazon-ecr-login`](https://github.com/aws-actions/amazon-ecr-login) in a prior job, which isn't available for all private registry implementations (works for ECR but that's about it), and you still want to take the same approach ideally without reinventing the wheel...

**Reply by sorliem** [reply]
Running into this exact issue with trying to access Google Artifact Registry with workload identity federation shortlived credentials

**Reply by cedws** [reply]
I was looking to do something similar. I wanted to generate a GitHub app token in an initial job configured with an Environment, so that only it could access the app private key, and then the short lived (1 hour) app token would be passed from the initial job to the main job. This way even if the job were compromised later down the line it would not be able to access the app private key, only the short lived app token.

**Comment by MiguelRipoll23** [comment]
I create dynamic secret using "gh secret set" with key/values of my variables and secrets to pass it to other workflow, it works but I wish there was a simpler way...

Edit: only works if you run the workflow two times when updating variables or secrets
